### PR TITLE
retry rather than block forever on pool checkout for unresponsive nodes

### PR DESCRIPTION
**Pull request by clockfort**

### CODE REVIEW

**Comment by tmgordeeva** [comment]
You'll get a NoSuchElementException from GenericObjectPool when you timeout so you won't discard that host.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolingContainer.java [hunk 1]
```diff
@@ -187,6 +187,7 @@ public Builder socketQueryTimeout(int val){
             config.setTimeBetweenEvictionRunsMillis(TimeUnit.MILLISECONDS.convert(1, TimeUnit.MINUTES));
             config.setNumTestsPerEvictionRun(-1); // Test all idle objects for eviction
             config.setJmxNamePrefix(host);
+            config.setMaxWaitMillis(socketTimeoutMillis);
```
Comment (schlosna) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolingContainer.java]: Should we be calling `clientPool.borrowObject(socketQueryTimeoutMillis)` in `runWithGoodResource`?
Do we need to retry up to some N times/backoff time limit if `borrowObject ` throws `NoSuchElementException`?

Comment (clockfort) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolingContainer.java]: borrowObject() just calls borrowObject(config.MaxWaitMillis) internally, so this is the same.

as per tanya's comment, yes, I will implement the retry for this exception.

Comment (schlosna) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolingContainer.java]: @clockfort sounds good, with retry we're probably good with the 2 second borrow timeout

**Pull request merged**